ext/pcntl/pcntl.c

```diff
@@ -1345,21 +1345,61 @@ static void pcntl_signal_handler(int signo)
 		/* oops, too many signals for us to track, so we'll forget about this one */
 		return;
 	}
-	PCNTL_G(spares) = psig->next;
 
-	psig->signo = signo;
-	psig->next = NULL;
+	struct php_pcntl_pending_signal *psig_first = psig;
+
+	/* Standard signals may be merged into a single one.
+	 * POSIX specifies that SIGCHLD has the si_pid field (https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/signal.h.html),
+	 * so we'll handle the merging for that signal. */
+	if (signo == SIGCHLD) {
+		/* Note: The first waitpid result is not necessarily the pid that was passed above!
+		 *       We therefore cannot avoid the first waitpid() call. */
+		int status;
+		pid_t pid;
+		while (true) {
+			do {
+				errno = 0;
+				pid = waitpid(WAIT_ANY, &status, WNOHANG);
```

Comment:
Linux docs say that `waitpid` with `WNOHANG` can't `EINTR`, but then again https://www.gnu.org/software/libc/manual/html_node/Merged-Signals.html looks like the code this is proposing, so it's probably better to keep it. Also not sure about other Unix systems.

Btw, might be good to link the above somewhere.

Reply:
About Linux docs: yes, but POSIX doesn't specify that so we should indeed keep it :) https://pubs.opengroup.org/onlinepubs/9699919799/functions/waitpid.html
And yes, forgot to link that, my mistake.